src/content/insertable-streams/webgpu/js/multi_video_main.js

```diff
@@ -163,43 +155,46 @@ class WebGPUTransform {
         });
     }
 
-    async renderOnScreen(videoFrame, gumFrame) {
+    async copyOnTexture(device, videoTexture, frame, xcorr, ycorr) {
+        if (!frame) {
+            return;
+        }
+        // Using GPUExternalTexture(when it's implemented for Breakout Box frames) will
+        // avoid making extra copies through ImageBitmap.
+        const videoBitmap = await createImageBitmap(frame, { resizeWidth: 480, resizeHeight: 270 });
+        device.queue.copyExternalImageToTexture(
+            { source: videoBitmap, origin: { x: 0, y: 0 } },
+            { texture: videoTexture, origin: { x: xcorr, y: ycorr } },
+            {
+                // the width of the image being copied
+                width: videoBitmap.width,
+                height: videoBitmap.height,
+            }
+        );
+        videoBitmap.close();
+        frame.close();
+    }
+
+    async renderOnScreen(videoSource, gumSource) {
         const device = this.device_;
-        if (device === null || device === undefined) {
+        const videoTexture = this.videoTexture_;
+        if (!device) {
             console.log('[WebGPUTransform] device is undefined or null.')
-            if (videoFrame) videoFrame.close();
-            if (gumFrame) gumFrame.close();
             return false;
         }
-        const videoTexture = this.videoTexture_;
-        let videoBitmap, gumBitmap;
-        if (videoFrame) {
-            videoBitmap = await createImageBitmap(videoFrame, { resizeWidth: 480, resizeHeight: 270 });
-            device.queue.copyExternalImageToTexture(
-                { source: videoBitmap, origin: { x: 0, y: 0 } },
-                { texture: videoTexture, origin: { x: 0, y: 270 } },
-                {
-                    // the width of the image being copied
-                    width: videoBitmap.width,
-                    height: videoBitmap.height,
-                }
-            );
-            videoBitmap.close();
-            videoFrame.close();
-        }
-        if (gumFrame) {
-            gumBitmap = await createImageBitmap(gumFrame, { resizeWidth: 480, resizeHeight: 270 });
-            device.queue.copyExternalImageToTexture(
-                { source: gumBitmap, origin: { x: 0, y: 0 } },
-                { texture: videoTexture, origin: { x: 480, y: 0 } },
-                {
-                    width: gumBitmap.width,
-                    height: gumBitmap.height,
-                }
-            );
-            gumBitmap.close();
-            gumFrame.close();
+
+        const videoPromise = videoSource.read().then(({ value }) => {
+            this.copyOnTexture(device, videoTexture, value, 0, 270);
+        });
+        const gumPromise = gumSource.read().then(({ value }) => {
+            this.copyOnTexture(device, videoTexture, value, 480, 0);
+        });
+        await videoPromise, gumPromise;
```

Comment:
I'm no expert on JS, but I think this is only awaiting videoPromise, not gumPromise. You probably want `await Promise.all(videoPromise, gumPromise);`.